client/merkle-mountain-range/src/offchain_mmr.rs

```diff
@@ -174,16 +200,58 @@ where
 				);
 			}
 		}
+		if self.best_canonicalized != header.number.saturating_sub(One::one()) {
+			warn!(
+				target: LOG_TARGET,
+				"Detected canonicalization skip: best {:?} current {:?}.",
+				self.best_canonicalized,
+				header.number,
+			);
+		}
+		self.best_canonicalized = header.number;
+	}
+
+	/// In case of missed finality notifications (node restarts for example),
+	/// make sure to also canon everything leading up to `notification.tree_route`.
+	pub fn canonicalize_catch_up(&mut self, notification: &FinalityNotification<B>) {
+		let first = notification.tree_route.first().unwrap_or(&notification.hash);
+		if let Some(mut header) = self.header_metadata_or_log(*first, "canonicalize") {
+			let mut to_canon = VecDeque::<<B as Block>::Hash>::new();
+			// Walk up the chain adding all blocks newer than `self.best_canonicalized`.
+			loop {
+				header = match self.header_metadata_or_log(header.parent, "canonicalize") {
+					Some(header) => header,
+					_ => break,
+				};
+				if header.number <= self.best_canonicalized {
+					break
+				}
+				to_canon.push_front(header.hash);
+			}
+			// Canonicalize all blocks leading up to current finality notification.
+			for hash in to_canon.drain(..) {
```

Comment:
Sorry for the dumb q, but is the canonicalization order that important here? If it isn't important then we can do it directly in previous loop without this temp`to_canon` vec. But maybe it breaks something in MMR structure and then this is required - just wanted to check

Reply:
we're saving `best_canonicalized` in the canonicalization function (with no checks, assuming block numbers always increasing), and also logging a warning when we see "skipped" blocks - that check is currently very simple: `if current != best.saturating_add(1) { warn!() }`;

the logic is kept simple in the canonicalization function with the assumption that blocks are canonicalized in order